### PR TITLE
DEPR: Deprecate cdate_range and merge into bdate_range

### LINKED ISSUE #17596: Merge cdate_range functionality into bdate_range
Following up on some discussion from #17554. 

The main point of discussion was regarding merging the functionality of `cdate_range` into `bdate_range`.  This would involve adding the two additional keywords supported by `cdate_range` to `bdate_range` (`weekmask` and `holidays`), which would only be used (if provided) when a custom frequency is passed to `bdate_range`.   

This wouldn't be terribly invasive, as `cdate_range` is only top-level on master (not 0.20.3), and there is no documentation mentioning it.  Performing the merge would help keep the top-level API clean, and reduce the number of `date_range` related functions. On the user end, the change would amount to `cdate_range(...)` -> `bdate_range(..., freq='C')`.

An alternative would be to merge both `bdate_range` and `cdate_range` into `date_range`.  The only difference between `bdate_range` and `date_range` is the default value of `freq`, so from a technical standpoint it wouldn't add complexity to implement compared to just merging `cdate_range` -> `bdate_range`.

This would be a more invasive change though, as `bdate_range` currently is currently top-level, and there is some documentation mentioning it, so there'd be some additional work in terms of deprecation/doc modification.  It would provide the advantage of having only one `date_range` related function.  On the user end, the change would amount to `cdate_range(...)` -> `date_range(..., freq='C')`, and `bdate_range(...)` -> `date_range(..., freq='B')`.

The current plan is to do the `cdate_range` -> `bdate_range` merge, but we're interested in community input.

cc @jorisvandenbossche

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #17596
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

A bit going on here, so see summary below.

Deprecated `cdate_range`:
- Removed from api (was added to api after 0.20.3 release)
- Added a `FutureWarning` to `cdate_range` indicating deprecation
    - Added a test for this too
- Converted other tests using `cdate_range` to use `bdate_range` instead
    - Cleaned up the entire test script to make it more consistent with others (i.e. `with` context)
- Removed a reference to `cdate_range` from the current whatsnew (previous commit I wrote)

Expanded functionality of `bdate_range`:
- Added the additional `cdate_range` params (`weekmask`, `holidays`) for custom frequency ranges
- Expanded that functionality beyond what was previously present in `cdate_range`
    - `cdate_range` only supported cday with the `weekmask` and `holidays` params, now all custom freqs are supported.
    - Did this by checking if the passed freq starts with 'C'.
    - Added a test that should catch if a new non-custom freq is added that starts with 'C'.
    - Didn't add this to `cdate_range`; left that as-is.
- Added a `UserWarning` if non-default `weekmask` or `holidays` are passed without a custom freq.
    - Added a test for this too.

Edits to `timeseries.rst`:
- Added a custom freq `bdate_range` example to the "Generating Ranges of Timestamps" section
- Reordered some of the examples in the "Generating Ranges of Timestamps" section
    - Tried to order examples by simplicity
- Otherwise, mostly small rewording changes and other such small fixes (ticks, capitalization, etc.).

### CODE REVIEW

File: doc/source/timeseries.rst [hunk 1]
```diff
@@ -76,21 +76,21 @@ Overview
...
+``DatetimeIndex``  Index of ``Timestamp``          ``to_datetime``, ``date_range``, ``bdate_range``, ``DatetimeIndex``
 ``Period``         Represents a single time span   ``Period``
 ``PeriodIndex``    Index of ``Period``             ``period_range``, ``PeriodIndex``
-=================  =============================== ==================================================
+=================  =============================== ===================================================================
 
 .. _timeseries.representation:
 
-Time Stamps vs. Time Spans
+Timestamps vs. Time Spans
```
Comment (jreback) [inline doc/source/timeseries.rst]: underlines should match the length exactly

File: doc/source/timeseries.rst [hunk 2]
```diff
@@ -238,44 +234,41 @@ Invalid Data
...
-for these is nanoseconds (since these are how ``Timestamp`` s are stored). However,
-often epochs are stored in another ``unit`` which can be specified. These are computed
-from the starting point specified by the :ref:`Origin Parameter <timeseries.origin>`.
-
-Typical epoch stored units
+pandas supports converting integer or float epoch times to ``Timestamp`` and 
+``DatetimeIndex``. The default unit is nanoseconds, since that is how ``Timestamp``
+objects are stored internally. However, epochs are often stored in another ``unit``
+which can be specified. These are computed from the starting point specified by the
+:ref:`Origin Parameter <timeseries.origin>`.
```
Comment (jreback) [inline doc/source/timeseries.rst]: use ``origin``

File: doc/source/timeseries.rst [hunk 3]
```diff
@@ -402,15 +392,29 @@ using various combinations of parameters like ``start``, ``end``,
 
    pd.bdate_range(start=start, periods=20)
 
-The start and end dates are strictly inclusive. So it will not generate any
```
Comment (jreback) [inline doc/source/timeseries.rst]: put a small ::warning box where you say ``cdate_range`` is deprecated, add a ref as well (then you can refer to it in whatsnew)

Comment (jreback) [reply doc/source/timeseries.rst]: in the appropriate location in this file

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -163,6 +163,7 @@ Other Enhancements
 - :func:`Categorical.rename_categories` now accepts a dict-like argument as `new_categories` and only updates the categories found in that dict. (:issue:`17336`)
 - :func:`read_excel` raises ``ImportError`` with a better message if ``xlrd`` is not installed. (:issue:`17613`)
 - :meth:`DataFrame.assign` will preserve the original order of ``**kwargs`` for Python 3.6+ users instead of sorting the column names
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: move to deprecation section, mention the deprecation, and include a ref

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -491,6 +492,7 @@ Deprecations
 - ``pd.options.html.border`` has been deprecated in favor of ``pd.options.display.html.border`` (:issue:`15793`).
 - :func:`SeriesGroupBy.nth` has deprecated ``True`` in favor of ``'all'`` for its kwarg ``dropna`` (:issue:`11038`).
 - :func:`DataFrame.as_blocks` is deprecated, as this is exposing the internal implementation (:issue:`17302`)
+- ``cdate_range`` has been deprecated in favor of :func:`bdate_range` (:issue:`17596`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: combine with the previous

File: pandas/core/indexes/datetimes.py [hunk 6]
```diff
@@ -2088,6 +2104,16 @@ def bdate_range(start=None, end=None, periods=None, freq='B', tz=None,
...
     """
 
+    if is_string_like(freq) and freq.startswith('C'):
+        try:
+            freq = prefix_mapping[freq](holidays=holidays, weekmask=weekmask)
+        except (KeyError, TypeError):
+            msg = 'invalid custom frequency string: {freq}'.format(freq=freq)
+            raise ValueError(msg)
+    elif holidays or (weekmask != 'Mon Tue Wed Thu Fri'):
+        warnings.warn('a custom frequency string was not passed, ignoring '
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: raise here, this is invalid

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 7]
```diff
@@ -525,20 +524,18 @@ def test_freq_divides_end_in_nanos(self):
 
 
 class TestCustomDateRange(object):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: separately, most of these classes could actually be a hierarchy instead to share some code (or could parametrize things), but separate / later PR.

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 8]
```diff
@@ -547,66 +544,92 @@ def test_cached_range(self):
...
+        # should not raise
+        bdate_range(START, END, freq=freq, weekmask='Mon Wed Fri',
+                    holidays=['2009-03-14'])
+
+        bad_freq = freq + 'FOO'
+        msg = 'invalid custom frequency string: {freq}'
+        with tm.assert_raises_regex(ValueError, msg.format(freq=bad_freq)):
+            bdate_range(START, END, freq=bad_freq)
+
+    def test_depr_cdaterange(self):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: deprecation

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good

File: doc/source/timeseries.rst [hunk 9]
```diff
@@ -238,44 +234,41 @@ Invalid Data
...
-Typical epoch stored units
+pandas supports converting integer or float epoch times to ``Timestamp`` and 
+``DatetimeIndex``. The default unit is nanoseconds, since that is how ``Timestamp``
+objects are stored internally. However, epochs are often stored in another ``unit``
+which can be specified. These are computed from the starting point specified by the
+:ref:`Origin Parameter <timeseries.origin>`.
 
 .. ipython:: python
 
+   pd.Timestamp(1349720105, unit='s')
```
Comment (jorisvandenbossche) [inline doc/source/timeseries.rst]: I personally don't find it necessary to add an example using `Timestamp`, IMO we should 'sell' `to_datetime` as the function to convert things to timestamps. (you can also convert a scalar with `to_datetime` if you want)

File: doc/source/timeseries.rst [hunk 10]
```diff
@@ -360,37 +351,36 @@ Index constructor and pass in a list of datetime objects:
    index = pd.Index(dates)
    index
 
-Practically, this becomes very cumbersome because we often need a very long
+In practice this becomes very cumbersome because we often need a very long
 index with a large number of timestamps. If we need timestamps on a regular
-frequency, we can use the pandas functions ``date_range`` and ``bdate_range``
-to create timestamp indexes.
+frequency, we can use the ``date_range`` and ``bdate_range`` functions
+to create a ``DatetimeIndex``. The default frequency for ``date_range`` is a
```
Comment (jorisvandenbossche) [inline doc/source/timeseries.rst]: can you make date_range and bdate_range references to their docstring? (``:func:`date_range` ``)

File: doc/source/timeseries.rst [hunk 11]
```diff
@@ -402,15 +392,29 @@ using various combinations of parameters like ``start``, ``end``,
 
    pd.bdate_range(start=start, periods=20)
 
-The start and end dates are strictly inclusive. So it will not generate any
-dates outside of those dates if specified.
+.. versionadded:: 0.21.0
+
+``bdate_range`` can also generate a range of custom frequency dates by using
```
Comment (jorisvandenbossche) [inline doc/source/timeseries.rst]: I think if you indent this paragraph, it more 'belongs' to the versionadded directive (although not fully sure how it will look like when rendered)

File: pandas/core/indexes/datetimes.py [hunk 12]
```diff
@@ -2049,7 +2051,8 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
 
 
 def bdate_range(start=None, end=None, periods=None, freq='B', tz=None,
-                normalize=True, name=None, closed=None, **kwargs):
+                normalize=True, name=None, weekmask='Mon Tue Wed Thu Fri',
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: I would put the weekmask default to None? (as it is only used when freq is something else, so this gives a bit a misleading signature?

And if `freq='C'` is passed and weekmask is None, you can set it to this value. 
(or is None a valid value when `freq='C'` ?)

File: pandas/core/indexes/datetimes.py [hunk 13]
```diff
@@ -2071,6 +2074,19 @@ def bdate_range(start=None, end=None, periods=None, freq='B', tz=None,
         Normalize start/end dates to midnight before generating date range
     name : string, default None
         Name of the resulting DatetimeIndex
+    weekmask : string, default 'Mon Tue Wed Thu Fri'
+        weekmask of valid business days, passed to ``numpy.busdaycalendar``,
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: weekmask -> Weekmask (also the description for 'holidays' can start with capital)

File: pandas/core/indexes/datetimes.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/datetimes.py]: I agree this can raise.

File: pandas/core/indexes/datetimes.py [hunk 14]
```diff
@@ -2137,6 +2165,9 @@ def cdate_range(start=None, end=None, periods=None, freq='C', tz=None,
     -------
     rng : DatetimeIndex
     """
+    warnings.warn("cdate_range is deprecated and will be removed in a future "
+                  "version, instead use bdate_range(..., freq='{freq}')"
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: bdate_range -> pd.bdate_range (to make it clear that bdate_range is top-level, as opposed to cdate_range)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot for this! 
Added a few comments

File: doc/source/timeseries.rst (hunk 11 above)
Comment (jschendel) [reply doc/source/timeseries.rst]: I tried indenting and it rendered everything on the same line, i.e. "_New in version 0.21.0:_ `bdate_range` can also...".  I changed it back to make it consistent with the other usages of versionadded within `timeseries.rst` and other documentation.

**Comment by jschendel** [comment]
Made the review changes and fixed some unrelated typos I noticed in whatsnew (missing : and `).

File: doc/source/timeseries.rst [hunk 15]
```diff
@@ -402,53 +394,84 @@ using various combinations of parameters like ``start``, ``end``,
...
-Since pandas represents timestamps in nanosecond resolution, the timespan that
+Since pandas represents timestamps in nanosecond resolution, the time span that
 can be represented using a 64-bit integer is limited to approximately 584 years:
 
 .. ipython:: python
 
    pd.Timestamp.min
    pd.Timestamp.max
 
-See :ref:`here <timeseries.oob>` for ways to represent data outside these bound.
```
Comment (jorisvandenbossche) [inline doc/source/timeseries.rst]: I think here the "for ways to represent data outside these bound" is actually useful to know that it is interesting to click this link

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

File: doc/source/timeseries.rst (hunk 15 above)
Comment (jorisvandenbossche) [reply doc/source/timeseries.rst]: Ah, but the 'see also' block will show the full title, which is "representing out of bounds spans", which also says it.

File: doc/source/timeseries.rst [hunk 16]
```diff
@@ -402,53 +394,84 @@ using various combinations of parameters like ``start``, ``end``,
...
+
+   weekmask = 'Mon Wed Fri'
+
+   holidays = [datetime(2011, 1, 5), datetime(2011, 3, 14)]
+
+   pd.bdate_range(start, end, freq='C', weekmask=weekmask, holidays=holidays)
+
+   pd.bdate_range(start, end, freq='CBMS', weekmask=weekmask)
+
+.. warning::
```
Comment (jreback) [inline doc/source/timeseries.rst]: move this to the top of the subsection. That way you can in the whatsnew refer directly to this section.

File: doc/source/whatsnew/v0.21.0.txt [hunk 17]
```diff
@@ -657,9 +658,9 @@ Numeric
 
 Categorical
 ^^^^^^^^^^^
-- Bug in :func:`Series.isin` when called with a categorical (:issue`16639`)
+- Bug in :func:`Series.isin` when called with a categorical (:issue:`16639`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: thanks for fixing this!

File: pandas/tests/api/test_api.py [hunk 18]
```diff
@@ -63,7 +63,7 @@ class TestPDApi(Base):
     # top-level functions
     funcs = ['bdate_range', 'concat', 'crosstab', 'cut',
              'date_range', 'interval_range', 'eval',
-             'factorize', 'get_dummies', 'cdate_range',
+             'factorize', 'get_dummies',
```
Comment (jreback) [inline pandas/tests/api/test_api.py]: move to the deprecated functions section, and add a test (in same module), model after the top-level deprecations.

File: pandas/tseries/offsets.py [hunk 19]
```diff
@@ -3002,5 +3003,3 @@ def generate_range(start=None, end=None, periods=None,
     FY5253,
     FY5253Quarter,
 ])
-
```
Comment (jreback) [inline pandas/tseries/offsets.py]: why did you change this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tseries/offsets.py (hunk 19 above)
Comment (jschendel) [reply pandas/tseries/offsets.py]: Cleaning.  Seemed a bit strange that `Nano` was being added to the dict after creation.  Looks like it was historically structured that way for compat with numpy < 1.7:

https://github.com/pandas-dev/pandas/blob/94e2194b7ab2b62126d306896ac42a8c00c4f41c/pandas/tseries/offsets.py#L2017-L2019

Structure was kept the same when the `if` statement was removed, but seems fine to directly add it to the dict now.

Comment (jreback) [reply pandas/tseries/offsets.py]: k cool

**Comment by jschendel** [comment]
Made the additional review changes.

File: pandas/tests/api/test_api.py [hunk 20]
```diff
@@ -240,3 +240,13 @@ def test_deprecation_access_func(self):
                 [c1, c2],
                 sort_categories=True,
                 ignore_order=True)
+
+
+class TestCDateRange(object):
+
+    def test_deprecation_cdaterange(self):
+        # GH17596
+        from pandas.core.indexes.datetimes import cdate_range
```
Comment (jreback) [inline pandas/tests/api/test_api.py]: import from pandas here (we are actually testing the pd. deprecation); in this case they are the same, but its our practice to use the one we are deprecating.

**Review by jreback — APPROVED** [review]
just a minor comment.

File: pandas/tests/api/test_api.py (hunk 20 above)
Comment (jorisvandenbossche) [reply pandas/tests/api/test_api.py]: no, we are testing this one (the move to top-level pd namespace was only very recently in master, so we don't have to deprecate that. That one we just removed, and we are deprecating the actual nested one)

Comment (jreback) [reply pandas/tests/api/test_api.py]: oh, ok then

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jschendel
